plutus-core/plutus-core/src/Bitwise.hs

```diff
@@ -0,0 +1,569 @@
+  -- Rotating by more than the number of bits in a ByteString 'wraps around',
+  -- so we're only interested in the rotation modulo the number of bits.
+  | otherwise = case i `rem` bitLen of
+            -- Means we have a multiple of the bit count, so nothing to do.
+            0         -> bs
+            magnitude -> overPtrLen bs $ \ptr len -> go ptr len magnitude
+  where
+    go :: Ptr Word8 -> Int -> Integer -> IO (Ptr Word8)
+    go src len displacement = do
+      dst <- mallocBytes len
+      case len of
+        -- If we only have one byte, we an borrow from the Bits instance for
+        -- Word8.
+        1 -> do
+          srcByte <- peek src
+          let srcByte' = srcByte `rotate` fromIntegral displacement
+          poke dst srcByte'
+        -- If we rotate by a multiple of 8, we only need to move around whole
+        -- bytes, rather than individual bits. Because we only move contiguous
+        -- blocks (regardless of rotation direction), we can do this using
+        -- memcpy, which is must faster, especially on larger ByteStrings.
+        _ -> case displacement `quotRem` 8 of
+          (bigMove, 0) -> do
+            let mainLen :: CSize = fromIntegral . abs $ bigMove
+            let restLen :: CSize = fromIntegral len - mainLen
+            void $ case signum bigMove of
+              1 -> memcpy (plusPtr dst . fromIntegral $ restLen) src mainLen >>
+                   memcpy dst (plusPtr src . fromIntegral $ mainLen) restLen
+              _ -> memcpy (plusPtr dst . fromIntegral $ mainLen) src restLen >>
+                   memcpy dst (plusPtr src . fromIntegral $ restLen) mainLen
+          -- If we don't rotate by a multiple of 8, we have to construct new
+          -- bytes, rather than just copying over old ones. We do this in two
+          -- steps:
+          --
+          -- 1. Compute the 'read offset' into the source ByteString based on
+          -- the rotation magnitude and direction.
+          -- 2. Use that read offset to perform an (unchecked) bit lookup for an
+          -- entire 8-bit block, then construct the byte that results.
+          --
+          -- We can do the bytes in the result in any order using this method:
+          -- we choose to do it in traversal order.
+          _ -> for_ [0 .. len - 1] $ \j -> do
+                let start = (len - 1 - j) * 8
+                let dstByte = foldl' (addBit start displacement) zeroBits [0 .. 7]
+                poke (plusPtr dst j) dstByte
+      pure dst
+    bitLen :: Integer
+    bitLen = fromIntegral $ BS.length bs * 8
+    addBit :: Int -> Integer -> Word8 -> Integer -> Word8
+    addBit start displacement acc offset =
+      let oldIx = (offset + fromIntegral start + bitLen - displacement) `rem` bitLen in
+        if dangerousRead bs oldIx
+        then acc .|. (bit . fromIntegral $ offset)
+        else acc
+
+{-# NOINLINE shiftByteString #-}
+shiftByteString :: ByteString -> Integer -> ByteString
```

Comment:
It'd be more costly, since I'd have to do the 'wraparound' logic of a rotate, _then_ a clear, which would involve extra writes. Furthermore, I'd still need a special case of 'shifting by more than the bit length', since doing a rotate in that case is _completely_ pointless.